go/consensus/tendermint/apps/registry/transactions.go

```diff
@@ -0,0 +1,452 @@
+package registry
+
+import (
+	"github.com/oasislabs/oasis-core/go/common/cbor"
+	"github.com/oasislabs/oasis-core/go/common/entity"
+	"github.com/oasislabs/oasis-core/go/common/node"
+	"github.com/oasislabs/oasis-core/go/consensus/tendermint/abci"
+	"github.com/oasislabs/oasis-core/go/consensus/tendermint/api"
+	registryState "github.com/oasislabs/oasis-core/go/consensus/tendermint/apps/registry/state"
+	stakingState "github.com/oasislabs/oasis-core/go/consensus/tendermint/apps/staking/state"
+	registry "github.com/oasislabs/oasis-core/go/registry/api"
+	staking "github.com/oasislabs/oasis-core/go/staking/api"
+)
+
+func (app *registryApplication) registerEntity(
+	ctx *abci.Context,
+	state *registryState.MutableState,
+	sigEnt *entity.SignedEntity,
+) error {
+	ent, err := registry.VerifyRegisterEntityArgs(app.logger, sigEnt, ctx.IsInitChain())
+	if err != nil {
+		return err
+	}
+
+	if ctx.IsCheckOnly() {
+		return nil
+	}
+
+	// Charge gas for this transaction.
+	params, err := state.ConsensusParameters()
+	if err != nil {
+		app.logger.Error("RegisterEntity: failed to fetch consensus parameters",
+			"err", err,
+		)
+		return err
+	}
+	if err = ctx.Gas().UseGas(1, registry.GasOpRegisterEntity, params.GasCosts); err != nil {
+		return err
+	}
+	if err = ctx.Gas().UseGas(len(ent.Nodes), registry.GasOpRegisterNode, params.GasCosts); err != nil {
+		return err
+	}
+
+	if !params.DebugBypassStake {
+		if err = stakingState.EnsureSufficientStake(ctx, ent.ID, []staking.ThresholdKind{staking.KindEntity}); err != nil {
+			app.logger.Error("RegisterEntity: Insufficent stake",
+				"err", err,
+				"id", ent.ID,
+			)
+			return err
+		}
+	}
+
+	// Make sure the signer of the transaction matches the signer of the entity.
+	if !ctx.IsInitChain() && !sigEnt.Signature.PublicKey.Equal(ctx.TxSigner()) {
```

Comment:
why is the check bypassed for InitChain()?

Reply:
Because in InitChain this method is not invoked from an actual transaction so there is no "transaction signer" and calling the `ctx.TxSigner()` method would panic.

Reply:
Will add a comment.